include/deal.II/lac/diagonal_matrix.h

```diff
@@ -0,0 +1,290 @@
+
+  /**
+   * Initialize with a given vector by copying the content of the vector
+   * @p vec.
+   */
+  void reinit (const VectorType &vec);
+
+  /**
+   * Returns a reference to the underlying vector for manipulation of the
+   * entries on the diagonal.
+   */
+  VectorType &get_vector();
+
+  /**
+   * Returns a read-only reference to the underlying vector.
+   */
+  const VectorType &get_vector() const;
+
+  /**
+   * Number of rows of this matrix. To remember: this matrix is an <i>m x
+   * m</i>-matrix.
+   */
+  size_type m () const;
+
+  /**
+   * Number of columns of this matrix. To remember: this matrix is an <i>n x
+   * n</i>-matrix.
+   */
+  size_type n () const;
+
+  /**
+   * Read-only access to a value. This is restricted to the case where
+   * <i>i==j</i>.
+   *
+   * If the underlying vector is distributed with MPI, not all of the indices
+   * <i>i</i> might actually be accessible. Refer to the method
+   * <code>get_vector().locally_owned_elements()</code> for the entries that
+   * actually are accessible.
+   */
+  value_type operator()(size_type i, size_type j) const;
+
+  /**
+   * Read-write access to a value. This is restricted to the case where
+   * <i>i==j</i>.
+   *
+   * If the underlying vector is distributed with MPI, not all of the indices
+   * <i>i</i> might actually be accessible. Refer to the method
+   * <code>get_vector().locally_owned_elements()</code> for the entries that
+   * actually are accessible.
+   */
+  value_type &operator()(size_type i, size_type j);
+
+  /**
+   * Add an array of values given by <tt>values</tt> in the given global
+   * matrix row at columns specified by col_indices. Due to the storage of
+   * this matrix, entries are only added to the diagonal of the matrix. All
+   * other entries are ignored and no exception is thrown.
+   *
+   * This function is for a consistent interface with the other matrix classes
+   * in deal.II and can be used in
+   * ConstraintMatrix::distribute_local_to_global to get exactly the same
+   * diagonal as when assembling into a sparse matrix.
+   */
+  template <typename number2>
+  void add (const size_type  row,
+            const size_type  n_cols,
+            const size_type *col_indices,
+            const number2   *values,
+            const bool       elide_zero_values = true,
+            const bool       col_indices_are_sorted = false);
+
+  /**
+   * Performs a matrix-vector multiplication with the given matrix.
+   */
+  void vmult (VectorType       &dst,
+              const VectorType &src) const;
+
+  /**
+   * Performs a transpose matrix-vector multiplication with the given
+   * matrix. Since this represents a diagonal matrix, exactly the same as
+   * vmult().
+   */
+  void Tvmult (VectorType       &dst,
+               const VectorType &src) const;
+
+  /**
+   * Adds the result of a matrix-vector multiplication into the destination
+   * vector dst. Needs to create a temporary vector, therefore it is not
```

Comment:
maybe have a mutable vector for this purpose so that you don't create it here every time, but only create it once in the constructor?